src/users/types.rs

```diff
@@ -0,0 +1,111 @@
+
+#[derive(Debug)]
+pub struct HumanUser {
+	/// The user's first name
+	pub first_name: String,
+	/// The user's last name
+	pub last_name: String,
+	/// The user's preferred username
+	pub preferred_username: String,
+	/// The user's email address
+	pub email: String,
+	/// The user's external ID
+	pub external_id: String,
+	/// The user's phone number
+	pub phone: Option<String>,
+	/// The user's organization id
+	pub organization_id: Option<String>,
+	/// Whether the user is enabled
+	pub enabled: bool,
+	/// The ID of the identity provider to link with, if any
+	pub idp_id: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(super) struct CreateHumanUserRequest {
+	pub(super) organization: Option<CreateHumanUserOrg>,
+	pub(super) profile: Profile,
+	pub(super) email: CreateHumanUserEmail,
+	pub(super) phone: Option<CreateHumanUserPhone>,
+	#[serde(rename = "idpLinks")]
+	pub(super) idp_links: Vec<Idp>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(super) struct CreateHumanUserEmail {
+	pub(super) email: String,
+	#[serde(rename = "isVerified")]
+	pub(super) is_verified: bool,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(super) struct CreateHumanUserPhone {
+	pub(super) phone: String,
+	#[serde(rename = "isVerified")]
+	pub(super) is_verified: bool,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(super) struct CreateHumanUserOrg {
+	#[serde(rename = "orgId")]
+	pub(super) org_id: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(super) struct Idp {
+	#[serde(rename = "idpId")]
+	idp_id: String,
+	#[serde(rename = "userId")]
+	user_id: String,
+	#[serde(rename = "userName")]
+	username: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(super) struct Profile {
+	#[serde(rename = "givenName")]
+	given_name: String,
+	#[serde(rename = "familyName")]
+	family_name: String,
+	#[serde(rename = "nickName")]
+	nick_name: String,
+	#[serde(rename = "displayName")]
+	display_name: String,
+}
+
+impl HumanUser {
+	fn display_name(&self) -> String {
+		format!("{} {}", self.first_name, self.last_name)
+	}
+	fn idp_links(&self) -> Vec<Idp> {
+		if let Some(idp_id) = self.idp_id.clone() {
+			vec![Idp { idp_id, user_id: self.external_id.clone(), username: self.display_name() }]
+		} else {
+			vec![]
+		}
+	}
+}
+
+impl From<HumanUser> for CreateHumanUserRequest {
+	fn from(value: HumanUser) -> Self {
+		Self {
+			organization: value
+				.organization_id
+				.as_ref()
+				.map(|org_id| CreateHumanUserOrg { org_id: org_id.clone() }),
+			profile: Profile {
+				given_name: value.first_name.clone(),
+				family_name: value.last_name.clone(),
+				nick_name: value.external_id.clone(),
+				display_name: value.display_name(),
+			},
+			email: CreateHumanUserEmail { email: value.email.clone(), is_verified: false },
+			phone: value
+				.phone
+				.as_ref()
+				.map(|phone| CreateHumanUserPhone { phone: phone.clone(), is_verified: false }),
+			idp_links: value.idp_links(),
+		}
+	}
```

Comment:
The idea here was to create our own representation of a user. I would like to also use this representation of the user when we search for a user. The `display_name` is build from `first_name` and `last_name`. The `is_verified` is set I'm changing based of the inputs of the `create_human_user` because I thought we don't need this info on the user struct. Do you think it's a good idea separate the internal request from the user representation?

Reply:
It's a generic library, I think we should expose zitadel api types as close as possible to what they really are except cases where the the types are super inconvenient. I don't think we should hide anything, this library can potentially be used outside of famedly where assumptions and requirements are totally different

Reply:
We may need those values internally too.

That said, I appreciate the attempt at making the type more ergonomic for rust purposes, deeply nested structs just to have two attributes linked to a type isn't particularly convenient.

I think the most convenient format for a rust library would be to provide a builder, where optional attributes can be set with function calls, so not everything has to be specified up-front. The attributes should closely match the optionalness of the upstream API, though.

Reply:
[`typed-builder`](https://crates.io/crates/typed-builder)